src/radix_connect/mobile/client.rs

```diff
@@ -1,49 +1,174 @@
+            secure_storage: WalletClientStorage::new(secure_storage),
+            new_sessions: RwLock::new(HashMap::new()),
+        }
+    }
+
+    #[uniffi::method]
+    pub async fn handle_linking_request(
+        &self,
+        request: RadixConnectMobileLinkRequest,
+    ) -> Result<Url> {
+        // 1. Get the handshake request from the relay service
+        let handshake_request = self
+            .relay_service
+            .get_session_handshake_request(request.session_id)
+            .await?;
+
+        // 2. Generate a new Diffie-Hellman key pair
+        let wallet_private_key = DiffieHellmanPrivateKey::generate()?;
+
+        // 3. Compute the shared secret
+        // random salt
+        let salt = hex_decode("000102030405060708090a0b0c0d0e0f").unwrap();
+        let info = hex_decode("f0f1f2f3f4f5f6f7f8f9").unwrap();
+        let encryption_key = wallet_private_key.hkdf_key_agreement(
+            &handshake_request.public_key,
+            &salt,
+            &info,
+        )?;
+
+        // 4. Create a new session
+        let session = Session::new(
+            request.session_id,
+            SessionOrigin::WebDapp(request.origin),
+            encryption_key,
+        );
+
+        {
+            self.new_sessions
+                .try_write()
+                .map(|mut new_sessions| {
+                    new_sessions.insert(request.session_id, session);
+                })
+                .unwrap();
+        }
+
+        // 5. TODO: use the actual dapp callback path
+        Ok(Url::from_str("https://example.com").unwrap())
+    }
+
+    #[uniffi::method]
+    pub async fn handle_dapp_interaction_request(
+        &self,
+        dapp_request: RadixConnectMobileDappRequest,
+    ) -> Result<RadixConnectMobileSessionRequest> {
+        let session;
+        {
+            let new_sessions = self.new_sessions.try_read().unwrap();
+
+            session = match new_sessions.get(&dapp_request.session_id) {
+                Some(session) => {
+                    // Save the session to the secure storage
+                    self.secure_storage.save_session(session.clone())?;
+                    let session_id = session.id;
+
+                    self.new_sessions
+                        .try_write()
+                        .map(|mut new_sessions| {
+                            new_sessions.remove(&session_id);
+                        })
+                        .unwrap();
+
+                    session.to_owned()
+                }
+                None => {
+                    self.session_from_secure_storage(dapp_request.session_id)?
+                }
+            };
+        }
+
+        self.relay_service
+            .get_wallet_interaction_requests(session)
+            .await?
+            .into_iter()
+            .find(|intraction| {
+                intraction.interaction_id == dapp_request.interaction_id
+            })
+            .map(|interaction| RadixConnectMobileSessionRequest {
```

Comment:
minor: I prefer to handle the bad case first, i.e. invert the last `map` and the `ok_or`, since that way the last statements **type** matches the return type of the fn, which I like.

Reply:
prefer `RadixConnectMobileSessionRequest::new` inside the map. kind of like we in Swift tend to have one single init all other calls, it is advantageous for debugging (and logging) to have one shared entry )